src/openms/source/QC/FragmentMassError.cpp

```diff
@@ -108,6 +108,105 @@ namespace OpenMS
     return theo_spectrum;
   }
 
+  void calculateFME_(PeptideIdentification& pep_id, const MSExperiment& exp, const QCBase::SpectraMap& map_to_spectrum, bool& print_warning, double tolerance, FragmentMassError::ToleranceUnit tolerance_unit, double& accumulator_ppm, UInt32& counter_ppm, WindowMower& window_mower_filter)
+  {
+    if (pep_id.getHits().empty())
+    {
+      OPENMS_LOG_WARN << "PeptideHits of PeptideIdentification with RT: " << pep_id.getRT() << " and MZ: " << pep_id.getMZ() << " is empty.";
+      return;
+    }
+
+    //---------------------------------------------------------------------
+    // FIND DATA FOR THEORETICAL SPECTRUM
+    //---------------------------------------------------------------------
+
+    // sequence
+    const AASequence& seq = pep_id.getHits()[0].getSequence();
+
+    // charge: re-calulated from masses since much more robust this way (PepID annotation of pep_id.getHits()[0].getCharge() could be wrong)
+    Int charge = static_cast<Int>(round(seq.getMonoWeight() / pep_id.getMZ()));
+
+    //-----------------------------------------------------------------------
+    // GET EXPERIMENTAL SPECTRUM MATCHING TO PEPTIDEIDENTIFICTION
+    //-----------------------------------------------------------------------
+
+    if (!pep_id.metaValueExists("spectrum_reference"))
+    {
+      throw Exception::InvalidParameter(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No spectrum reference annotated at peptide identifiction!");
+    }
+    const MSSpectrum& exp_spectrum = exp[map_to_spectrum.at(pep_id.getMetaValue("spectrum_reference").toString())];
+
+    if (exp_spectrum.getMSLevel() != 2)
+    {
+      throw Exception::IllegalArgument(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "The matching spectrum of the mzML is not an MS2 Spectrum.");
+    }
+    Precursor::ActivationMethod act_method;
+    if (exp_spectrum.getPrecursors().empty() || exp_spectrum.getPrecursors()[0].getActivationMethods().empty())
+    {
+      if (print_warning) OPENMS_LOG_WARN << "No MS2 activation method provided. Using CID as fallback to compute fragment mass errors." << std::endl;
+      print_warning = false; // only print it once
+      act_method = Precursor::ActivationMethod::CID;
+    }
+    else
+    {
+      act_method = *exp_spectrum.getPrecursors()[0].getActivationMethods().begin();
+    }
+
+    //---------------------------------------------------------------------
+    // CREATE THEORETICAL SPECTRUM
+    //---------------------------------------------------------------------
+    PeakSpectrum theo_spectrum = getTheoSpec_(act_method, seq, charge);
+
+    //-----------------------------------------------------------------------
+    // COMPARE THEORETICAL AND EXPERIMENTAL SPECTRUM
+    //-----------------------------------------------------------------------
+    if (exp_spectrum.empty() || theo_spectrum.empty())
```

Comment:
does theo_spectrum.empty() need to be tested? are there corner cases you encountered where this happened?

Reply:
same here